src/lua/popen.c

```diff
@@ -1642,38 +1643,58 @@ lbox_popen_kill(struct lua_State *L)
  * - FiberIsCancelled: cancelled by an outside code.
  *
  * Return a process status table (the same as ph.status and
- * ph.info().status). @see lbox_popen_info() for the format
- * of the table.
+ * ph.info().status) at success. @see lbox_popen_info() for
+ * the format of the table.
+ *
+ * Return `nil, err` on a timeout:
+ *
+ * - TimedOut: @a timeout quota is exceeded.
  */
 static int
 lbox_popen_wait(struct lua_State *L)
 {
-	/*
-	 * FIXME: Use trigger or fiber conds to sleep and wake up.
-	 * FIXME: Add timeout option: ph:wait({timeout = <...>})
-	 */
-	struct popen_handle *handle;
+	/* Extract handle. */
 	bool is_closed;
-	if ((handle = luaT_check_popen_handle(L, 1, &is_closed)) == NULL) {
-		diag_set(IllegalParams, "Bad params, use: ph:wait()");
-		return luaT_error(L);
-	}
+	struct popen_handle *handle = luaT_check_popen_handle(L, 1, &is_closed);
+	if (handle == NULL)
+		goto usage;
 	if (is_closed)
 		return luaT_popen_handle_closed_error(L);
 
-	int state;
-	int exit_code;
+	ev_tstamp timeout = TIMEOUT_INFINITY;
 
-	while (true) {
-		popen_state(handle, &state, &exit_code);
-		assert(state < POPEN_STATE_MAX);
-		if (state != POPEN_STATE_ALIVE)
-			break;
-		fiber_sleep(POPEN_LUA_WAIT_DELAY);
-		luaL_testcancel(L);
+	/* Extract options. */
+	if (!lua_isnoneornil(L, 2)) {
+		if (lua_type(L, 2) != LUA_TTABLE)
+			goto usage;
+
+		lua_getfield(L, 2, "timeout");
+		if (!lua_isnil(L, -1))
+			timeout = luaT_check_timeout(L, -1);
+		if (timeout < 0.0)
+			goto usage;
+		lua_pop(L, 1);
+	}
+
+	int rc = popen_wait_timeout(handle, timeout);
+	if (rc < 0) {
+		struct error *e = diag_last_error(diag_get());
+		if (e->type == &type_FiberIsCancelled)
+			return luaT_error(L);
```

Comment:
Nit: Better use `luaL_testcancel()` here instead of checking the error type.

Reply:
My personal opinion is that errors *are* for checking their types :)

In object oriented languages `try-catch` (C++, Java), `try-except` (Python) blocks have a syntax for handling errors of different types differently in a easy way. Looking at an error type is the usual way to make programmatic decisions (retry, ignore, report).

Our error related functions allow this, but, as you see, it looks a bit wordy. I don't think that it is a big problem per se. (The lack of focus on a well defined error hierarchy and lack of systematic attention to errors as part a function's contract is the problem, as for me.)

I like how the logic looks when we define function's exit states via formal definition of exit states of functions we call. In this case:

| `popen_wait_timeout()` | -> | `<popen handle>:wait()` |
| ---------------------- | -- | ----------------------- |
| success                | -> | return process_status   |
| TimedOut               | -> | return nil, TimedOut    |
| FiberIsCancelled       | -> | raise FiberIsCancelled  |

It is a direct mapping. If we'll use `luaL_testcancel()`, the control flow graph will look so:

| `popen_wait_timeout()` | -> | `luaL_testcancel()`                  | -> | `<popen handle>:wait()` |
| ---------------------- | -- | ------------------------------------ | -- | ----------------------- |
| success                | -> | if cancelled, raise FiberIsCancelled | -> | return process_status   |
| TimedOut               | -> | if cancelled, raise FiberIsCancelled | -> | return nil, TimedOut    |
| FiberIsCancelled       | -> | if cancelled, raise FiberIsCancelled | -> | unreachable  |

The code will be a bit more concise, but it'll contain implicit assumptions about relations between fiber's state and errors from a lower level functions.

I would prefer checking of an error type (despite that it requires one or two more lines of code).

Reply:
> My personal opinion is that errors _are_ for checking their types :)

Branching depending on an error/exception type isn't just ugly, but also makes the code flow look indirect. IMO error types are needed for logging only.

> The code will be a bit more concise, but it'll contain implicit assumptions about relations between fiber's state and errors from a lower level functions.

If a fiber was cancelled, there's no point to return `nil, error` to it, because it isn't functional and going to die pretty soon anyway.

What I personally like about `luaL_testcancel` is that it's easy to grep cancellation points in the Lua C code. Besides, we usually stick to calling `luaL_testcancel` instead of checking the exception type. In fact, the only place in the code where we do check exception types is ... `popen.c`. Well, looks like your code is consistent with the existing code after all :)

Feel free to leave as is.

Reply:
> Branching depending on an error/exception type isn't just ugly, but also makes the code flow look indirect. IMO error types are needed for logging only.

Hm. For me it looks more direct this way. At least, it is kinda mapping from a low level function contract to the new contract.

> If a fiber was cancelled, there's no point to return nil, error to it, because it isn't functional and going to die pretty soon anyway.

It is raised in the current code (using `luaT_error()`). All popen code follow the following principle: raise IllegalParams and FiberIsCancelled, return all other errors as `nil, err`. (It was a kind of official position of the team: report errors using `nil, err` with exception of parameters validation and fiber cancellation. If I would implement popen now, I would consider raising all errors.)

BTW, some kinds of failures assume that a user likely will need to check the error type, at least TimedOut in `:read()` and `:write()`.

> What I personally like about `luaL_testcancel` is that it's easy to grep cancellation points in the Lua C code.

Good point.

Regarding how popen and other code looks: I know. I would also note that popen is the only built-in Lua module (AFAIK) that has formal description of all exit situations and corresponding error types.

Thank you for all your points, I'll look on my code once again in reflection of them.

Reply:
I left it as is. Don't know, to be honest. It seems, it is hard for me to look from your perspective. Maybe other reviewers will add something that would change my mind.